### PR TITLE
Block height discrepancy

### PR DESCRIPTION
**Pull request by orbitalturtle**
#### Pull Request Checklist

- [ ] If this is your first time contributing, we recommend you read the [Code
  Contribution Guidelines](https://github.com/lightningnetwork/lnd/blob/master/docs/code_contribution_guidelines.md)
- [ ] All changes are Go version 1.12 compliant
- [ ] The code being submitted is commented according to [Code Documentation and Commenting](https://github.com/lightningnetwork/lnd/blob/master/docs/code_contribution_guidelines.md#CodeDocumentation)
- [ ] For new code: Code is accompanied by tests which exercise both
  the positive and negative (error paths) conditions (if applicable)
- [ ] For bug fixes: Code is accompanied by new tests which trigger
  the bug being fixed to prevent regressions
- [ ] Any new logging statements use an appropriate subsystem and
  logging level
- [ ] Code has been formatted with `go fmt`
- [ ] Protobuf files (`lnrpc/**/*.proto`) have been formatted with
  `make rpc-format` and compiled with `make rpc`
- [ ] New configuration flags have been added to `sample-lnd.conf`
- [ ] For code and documentation: lines are wrapped at 80 characters
  (the tab character should be counted as 8 characters, not 4, as some IDEs do
  per default)
- [ ] Running `make check` does not fail any tests
- [ ] Running `go vet` does not report any issues
- [ ] Running `make lint` does not report any **new** issues that did not
  already exist
- [ ] All commits build properly and pass tests. Only in exceptional
  cases it can be justifiable to violate this condition. In that case, the
  reason should be stated in the commit message.
- [ ] Commits have a logical structure according to [Ideal Git Commit Structure](https://github.com/lightningnetwork/lnd/blob/master/docs/code_contribution_guidelines.md#IdealGitCommitStructure)

### CODE REVIEW

File: routing/router.go [hunk 1]
```diff
@@ -1045,60 +1052,64 @@ func (r *ChannelRouter) networkHandler() {
...
+                        	         		log.Errorf("unable to filter block")
+						}
+
+						err = r.updateGraphWithClosedChannels(filteredBlock, uint32(missingBlock.Height))
+						if err != nil {
+                        			        log.Errorf("unable to prune graph with closed channels: "+
+                        			                "%v", err)
+                        			}
+					}
+			        } else {
```
Comment (gkrizek) [inline routing/router.go:1102]: I would get rid of the `else {continue}` and add an `else if chainUpdate.Height < currentHeight+1 { `

If we hit a scenario where it falls back to a lesser height than expected we could just log a message like

```
latest block height was already processed. skipping channel pruning.
```

Then have a `continue` under the log. Feel free to update that message though, haha.

**Review by gkrizek — COMMENTED** [review]
@orbitalturtle Looks good at first pass! Just some recommendations on syntax more than anything. I'll try to test this out soon on some real nodes.

File: routing/router.go [hunk 2]
```diff
@@ -1045,60 +1052,64 @@ func (r *ChannelRouter) networkHandler() {
 
 			// We'll ensure that any new blocks received attach
 			// directly to the end of our main chain. If not, then
-			// we've somehow missed some blocks. We don't process
-			// this block as otherwise, we may miss on-chain
-			// events.
+			// we've somehow missed some blocks. Here we'll catch 
+			// up the chain with the latest blocks.
 			currentHeight := atomic.LoadUint32(&r.bestHeight)
 			if chainUpdate.Height != currentHeight+1 {
```
[the remark this answers is not in the document]
Comment (orbitalturtle) [reply routing/router.go:1058]: good point

**Comment by orbitalturtle** [comment]
@gkrizek Awesome, just pushed up those changes. Excited to see how it works on the real nodes!

**Comment by gkrizek** [comment]
Ok finally got around to testing this on real btcd nodes. I don't think it quite worked like we wanted to. I had two btcd nodes and let one of them fall behind while the other was up to date. There was probably 6 or 8 blocks between them. I failed over from behind node to the synced node. I saw the `out of order blocks` message, but nothing after that. I expected to see the `Pruning channel graph using block` message that's inside of `updateGraphWithClosedChannels`. I haven't actually seen any messages from the `CRTR` subserved  since the out of order blocks message. So I think something is off there. 

I wanted to give you a heads up but I'll also try putting in a ton of debug statements and trying to again so we know exactly what's going on. Here's also the log file from the experiment:

```
2021-03-30 03:58:16.420 [INF] UTXN: Attempting to graduate height=1968614: num_kids=0, num_babies=0
2021-03-30 03:58:16.420 [INF] NTFN: New block: height=1968614, sha=00000000000000160feaa27d96ca11f2616fa5f8acf618ac26f07bcb9e2a4806
2021-03-30 03:57:46.735 [INF] DISC: Broadcasting 5 new announcements in 1 sub batches
2021-03-30 03:57:12.295 [INF] UTXN: Attempting to graduate height=1968613: num_kids=0, num_babies=0
2021-03-30 03:57:12.295 [INF] NTFN: New block: height=1968613, sha=0000000000000033978905cc1cbb5bd1c32db86e0d62c917b2fed6fa28e04adc
2021-03-30 03:57:12.293 [INF] UTXN: Attempting to graduate height=1968612: num_kids=0, num_babies=0
2021-03-30 03:57:12.293 [INF] NTFN: New block: height=1968612, sha=0000000000000017447a1d579110bad12db913851f201b9f6d699f2c1630ebfb
2021-03-30 03:57:12.291 [INF] UTXN: Attempting to graduate height=1968611: num_kids=0, num_babies=0
2021-03-30 03:57:12.291 [INF] NTFN: New block: height=1968611, sha=00000000000000379467bbb3f95a766603a1c9f58fbf04d1a6488158bd94d6aa
2021-03-30 03:57:12.288 [INF] UTXN: Attempting to graduate height=1968610: num_kids=0, num_babies=0
2021-03-30 03:57:12.288 [INF] NTFN: New block: height=1968610, sha=000000000000002089a194d77c4916bb3334515e776d8ff26a1901734975fa0b
2021-03-30 03:57:12.285 [INF] UTXN: Attempting to graduate height=1968609: num_kids=0, num_babies=0
2021-03-30 03:57:12.285 [INF] NTFN: New block: height=1968609, sha=0000000000000021fed4547830aa02b2f56b547657507dca9c2c03423c8db392
2021-03-30 03:57:12.273 [INF] NTFN: Missed blocks, attempting to catch up
2021-03-30 03:57:12.271 [ERR] CRTR: out of order block: expecting height=1968609, got height=1968613
2021-03-30 03:56:40.644 [INF] CRTR: Processed channels=0 updates=5 nodes=0 in last 59.999965892s
2021-03-30 03:56:16.735 [INF] DISC: Broadcasting 2 new announcements in 1 sub batches
2021-03-30 03:56:04.523 [INF] LNWL: Finished rescan for 0 addresses (synced to block 0000000000000017447a1d579110bad12db913851f201b9f6d699f2c1630ebfb, height 1968612)
2021-03-30 03:56:04.523 [INF] LNWL: Done catching up block hashes
2021-03-30 03:56:04.511 [INF] LNWL: Catching up block hashes to height 1968612, this might take a while
2021-03-30 03:56:04.509 [INF] LNWL: Started rescan from block 00000000000000162446f0c5cb36f0b1122c123332a97046bc6777817f8953a9 (height 1968608) for 0 addresses
2021-03-30 03:55:40.644 [INF] CRTR: Processed channels=0 updates=2 nodes=0 in last 59.999925677s
2021-03-30 03:54:46.735 [INF] DISC: Broadcasting 28 new announcements in 3 sub batches
2021-03-30 03:54:40.644 [INF] CRTR: Processed channels=0 updates=28 nodes=0 in last 1m29.958148064s
2021-03-30 03:53:16.735 [INF] DISC: Broadcasting 3 new announcements in 1 sub batches
2021-03-30 03:52:10.685 [INF] CRTR: Processed channels=0 updates=1 nodes=2 in last 59.999921867s
2021-03-30 03:51:46.735 [INF] DISC: Broadcasting 1 new announcements in 1 sub batches
2021-03-30 03:51:10.685 [INF] CRTR: Processed channels=0 updates=1 nodes=0 in last 59.999839273s
2021-03-30 03:50:16.735 [INF] DISC: Broadcasting 14 new announcements in 2 sub batches
2021-03-30 03:50:10.686 [INF] CRTR: Processed channels=0 updates=14 nodes=0 in last 1m15.045334968s
2021-03-30 03:48:46.735 [INF] DISC: Broadcasting 20 new announcements in 2 sub batches
2021-03-30 03:47:55.640 [INF] CRTR: Processed channels=0 updates=20 nodes=0 in last 59.999921204s
2021-03-30 03:47:16.735 [INF] DISC: Broadcasting 14 new announcements in 2 sub batches
2021-03-30 03:46:55.640 [INF] CRTR: Processed channels=0 updates=14 nodes=0 in last 59.999971604s
2021-03-30 03:45:55.640 [INF] CRTR: Processed channels=0 updates=13 nodes=0 in last 3m8.730284786s
2021-03-30 03:45:46.735 [INF] DISC: Broadcasting 10 new announcements in 1 sub batches
2021-03-30 03:44:16.735 [INF] DISC: Broadcasting 3 new announcements in 1 sub batches
2021-03-30 03:43:20.700 [INF] DISC: GossipSyncer(025f8e7cd929a95841664e756fe7ea105a3831eacb52ed5a07e70f7b472e023fda): fetching chan anns for 43 chans
2021-03-30 03:43:19.178 [INF] DISC: GossipSyncer(025f8e7cd929a95841664e756fe7ea105a3831eacb52ed5a07e70f7b472e023fda): sending final chan range chunk, size=2779
2021-03-30 03:43:19.176 [INF] DISC: GossipSyncer(025f8e7cd929a95841664e756fe7ea105a3831eacb52ed5a07e70f7b472e023fda): sending range chunk of size=8000
2021-03-30 03:43:19.170 [INF] DISC: GossipSyncer(025f8e7cd929a95841664e756fe7ea105a3831eacb52ed5a07e70f7b472e023fda): filtering chan range: start_height=0, num_blocks=1968608
2021-03-30 03:42:46.735 [INF] DISC: Broadcasting 1 new announcements in 1 sub batches
2021-03-30 03:41:46.910 [INF] CRTR: Processed channels=0 updates=1 nodes=0 in last 59.999927386s
2021-03-30 03:41:16.735 [INF] DISC: Broadcasting 31 new announcements in 4 sub batches
2021-03-30 03:40:46.910 [INF] CRTR: Processed channels=1 updates=29 nodes=1 in last 59.999952621s
2021-03-30 03:39:46.910 [INF] CRTR: Processed channels=0 updates=3 nodes=0 in last 2m11.113721094s
2021-03-30 03:39:46.735 [INF] DISC: Broadcasting 3 new announcements in 1 sub batches
2021-03-30 03:39:35.978 [INF] CRTR: Block 00000000000000162446f0c5cb36f0b1122c123332a97046bc6777817f8953a9 (height=1968608) closed 0 channels
2021-03-30 03:39:35.974 [INF] UTXN: Attempting to graduate height=1968608: num_kids=0, num_babies=0
2021-03-30 03:39:35.974 [INF] NTFN: New block: height=1968608, sha=00000000000000162446f0c5cb36f0b1122c123332a97046bc6777817f8953a9
2021-03-30 03:39:35.970 [INF] CRTR: Pruning channel graph using block 00000000000000162446f0c5cb36f0b1122c123332a97046bc6777817f8953a9 (height=1968608)
2021-03-30 03:38:46.910 [INF] CRTR: Block 000000000000003e7dd1e955da74ced6b7430cb5cbdb29e5295821aca676746b (height=1968607) closed 1 channels
2021-03-30 03:38:46.907 [INF] CHDB: Pruned 2 unconnected nodes from the channel graph
2021-03-30 03:38:46.907 [INF] CHDB: Pruned unconnected node 0246892404745f6233724a94e1f23fb73980e7937cc785afc65e5100441d994a7e from channel graph
2021-03-30 03:38:46.907 [INF] CHDB: Pruned unconnected node 032d8e70c02c4e307ad157cd5c30f89f75b06de2a7855e98557ff49babbc5e6c37 from channel graph
2021-03-30 03:38:46.903 [INF] UTXN: Attempting to graduate height=1968607: num_kids=0, num_babies=0
2021-03-30 03:38:46.903 [INF] NTFN: New block: height=1968607, sha=000000000000003e7dd1e955da74ced6b7430cb5cbdb29e5295821aca676746b
```

**Comment by orbitalturtle** [comment]
Dang! Lol. Okay I'll give this a second look

File: routing/router.go [hunk 3]
```diff
@@ -1045,60 +1052,63 @@ func (r *ChannelRouter) networkHandler() {
...
-			// closed channels.
-			closeSummaries := createCloseSummaries(blockHeight, chansClosed...)
-			r.notifyTopologyChange(&TopologyChange{
-				ClosedChannels: closeSummaries,
-			})
+					// We'll walk through all the outdated blocks
+					// and make sure we're able to update the graph
+					// with any closed channels from them.
+					for i := 0; i < blockDifference; i++ {
+						missingBlock := <-epochClient.Epochs
```
Comment (gkrizek) [inline routing/router.go:1095]: It seems like this is the problem. I added a bunch of logs to find where the catch is and I think it gets stuck here. Everything runs as expected until it hits here and it just gets stuck. Never continues..

Comment (orbitalturtle) [reply routing/router.go:1095]: Ahhh interesting. Thanks for finding that out. I've been unable to replicate the error for testing beyond the unit test so I feel kinda useless lol. It seems like the epochClient isn't receiving the epochs for some reason, so the program just blocks. I'll fiddle around with it ~ there must be something slightly different going on in the unit test that's getting it to pass this point

By chance, were you able to see if the program got inside the if statement right above it?

Comment (gkrizek) [reply routing/router.go:1095]: Are you referring to the `if epochClient == nil {` block? I forgot to add log statements in there. I can confirm it doesn't hit that error in there and I can conform that we at least start the `for` loop and when processing the first iteration that's when it gets stuck.

Comment (orbitalturtle) [reply routing/router.go:1095]: Ok cool, I think I figured out the problem. I'll push up the fix tomorrow. Sorry about that!

**Comment by orbitalturtle** [comment]
@gkrizek Pushed up the fix to the bug so it doesn't get stuck anymore! Seems it was a pointer-related issue

**Comment by gkrizek** [comment]
Awesome! Thank you. I'll try to test soon

**Pull request merged**